### PR TITLE
feat(kit): `Tiles` add component

### PR DESCRIPTION
**Pull request by waterplea**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Closes # 

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: projects/cdk/utils/miscellaneous/array-equals.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export function tuiArrayEquals<T>(a: readonly T[], b: readonly T[]): boolean {
```
Comment (splincode) [inline projects/cdk/utils/miscellaneous/array-equals.ts:1]: maybe `tuiArrayShallowEquals`?

Comment (splincode) [inline projects/cdk/utils/miscellaneous/array-equals.ts:1]: What do you think about `tuiArrayShallowEquals`?

**Review by splincode — APPROVED** [review]

File: projects/kit/components/tiles/tiles.component.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        const elements = Array.from(this.elementRef.nativeElement.children);
+        const currentIndex = elements.indexOf(this.element);
+        const newIndex = elements.indexOf(element);
+        const items = [...this.items];
+        const dragged = items[currentIndex];
+
+        items[currentIndex] = items[newIndex];
+        items[newIndex] = dragged;
+
+        this.itemsChange.emit(items);
```
[the remark this answers is not in the document]
Comment (waterplea) [reply projects/kit/components/tiles/tiles.component.ts:62]: We can add it later if there's a need for it.

File: projects/kit/components/tiles/tile-handle.directive.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    onPointer(x = NaN, y = NaN): void {
+        this.x = x - this.tile.element.offsetLeft;
+        this.y = y - this.tile.element.offsetTop;
+        this.tile.onDrag(!Number.isNaN(x));
+        this.tile.offset$.next([0, 0]);
+    }
+
+    @HostListener(`document:pointermove.silent`, [`$event.x`, `$event.y`])
+    onMove(x: number, y: number): void {
+        if (!Number.isNaN(this.x)) this.tile.offset$.next([x - this.x, y - this.y]);
```
Comment (splincode) [inline projects/kit/components/tiles/tile-handle.directive.ts:36]: please, rebase with main branch

**Review by nsbarsukov — APPROVED** [review]

File: projects/cdk/decorators/debounce.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+export function tuiDebounce(timeout: number): MethodDecorator {
+    let timeoutRef: ReturnType<typeof setTimeout> | undefined;
+
+    return function (_target, _key: string | symbol, descriptor: PropertyDescriptor) {
+        const {value} = descriptor;
+
+        descriptor.value = function (...args: unknown[]) {
+            clearTimeout(timeoutRef);
```
[the remark this answers is not in the document]
Comment (waterplea) [reply projects/cdk/decorators/debounce.ts:8]: `inject` cannot be used outside constructor context. There is still no way to legally access DI inside decorators.

**Review by vladimirpotekhin — APPROVED** [review]

File: projects/cdk/decorators/debounce.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (splincode) [reply projects/cdk/decorators/debounce.ts:8]: what do you think about

```ts
descriptor.value = function (...args: unknown[]) { 
    const ngZone = target.ngZone; 
    const windowRef = target.windowRef; 

    ngZone.runOutsideAngular(() => { 
         windowRef.clearTimeout(timeoutRef); 
         timeoutRef = windowRef.setTimeout(() => value.apply(this, args), timeout); });
     }
```

```ts
  constructor(
     @Inject(ElementRef) private readonly elementRef: ElementRef<Element>,
     @Inject(NgZone) readonly ngZone: NgZone,
     @Inject(WINDOW) readonly windowRef: Window,
   ) {}

    @tuiDebounce(500)
    @HostListener(`pointerleave`)
    reorder(element?: Element): void {
```

**Pull request merged**